controllers/datasource_controller.go

```diff
@@ -166,151 +166,200 @@ func (r *GrafanaDatasourceReconciler) Reconcile(ctx context.Context, req ctrl.Re
+				pluginErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
 			}
 		}
 
 		// then import the datasource into the matching grafana instances
 		err = r.onDatasourceCreated(ctx, &grafana, cr, datasource, hash)
 		if err != nil {
-			success = false
-			cr.Status.LastMessage = err.Error()
-			controllerLog.Error(err, "error reconciling datasource", "datasource", cr.Name, "grafana", grafana.Name)
+			applyErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
 		}
 	}
 
-	// if the datasource was successfully synced in all instances, wait for its re-sync period
-	if success {
-		cr.Status.LastMessage = ""
-		cr.Status.Hash = hash
-		if cr.ResyncPeriodHasElapsed() {
-			cr.Status.LastResync = metav1.Time{Time: time.Now()}
-		}
-		cr.Status.UID = cr.CustomUIDOrUID()
-		return ctrl.Result{RequeueAfter: cr.Spec.ResyncPeriod.Duration}, r.Client.Status().Update(ctx, cr)
-	} else {
-		// if there was an issue with the datasource, update the status
-		return ctrl.Result{RequeueAfter: RequeueDelay}, r.Client.Status().Update(ctx, cr)
+	// NOTE New Condition?
+	// Specific to datasources
+	if len(pluginErrors) > 0 {
+		err := fmt.Errorf("%v", pluginErrors)
+		r.Log.Error(err, "failed to apply plugins to all instances")
+	}
+
+	if len(applyErrors) > 0 {
+		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
 	}
+
+	condition := buildSynchronizedCondition("Datasource", conditionDatasourceSynchronized, cr.Generation, applyErrors, len(instances))
+	meta.SetStatusCondition(&cr.Status.Conditions, condition)
+
+	cr.Status.Hash = hash
+	cr.Status.LastMessage = "" // nolint:staticcheck
+	cr.Status.UID = cr.CustomUIDOrUID()
+
+	return ctrl.Result{RequeueAfter: cr.Spec.ResyncPeriod.Duration}, nil
 }
 
-func (r *GrafanaDatasourceReconciler) onDatasourceDeleted(ctx context.Context, namespace string, name string) error {
-	list := v1beta1.GrafanaList{}
-	opts := []client.ListOption{}
-	err := r.Client.List(ctx, &list, opts...)
+func (r *GrafanaDatasourceReconciler) deleteOldDatasource(ctx context.Context, cr *v1beta1.GrafanaDatasource) error {
+	instances, err := GetScopedMatchingInstances(r.Log, ctx, r.Client, cr)
 	if err != nil {
-		return err
+		return fmt.Errorf("fetching instances: %w", err)
 	}
 
-	for _, grafana := range list.Items {
+	for _, grafana := range instances {
 		grafana := grafana
-		if found, uid := grafana.Status.Datasources.Find(namespace, name); found {
-			grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, &grafana)
-			if err != nil {
+
+		found, uid := grafana.Status.Datasources.Find(cr.Namespace, cr.Name)
+		if !found {
+			continue
+		}
+
+		grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, &grafana)
+		if err != nil {
+			return err
+		}
+
+		datasource, err := grafanaClient.Datasources.GetDataSourceByUID(*uid)
+		if err != nil {
```

Comment:
I think we can have the conditional a bit cleaner by checking for the not found error first

```suggestion
		if errors.As(err,&notFound) {
		  continue
```